_posts/2022-02-04-shard-indexing-backpressure-in-opensearch.markdown

```diff
@@ -0,0 +1,205 @@
+
+Primary parameters, are leading indicators for node duress and are governed through soft limits. We have two such primary parameters:
+
+1. **Shard memory limit breach:** The shard memory limit is breached if the allocated shard level memory limit is exceeded and there is a need to increase the allocation. At any given point the memory distribution algorithm will try to maintain that the current memory utilization of shard approximately 85% of the assigned shard memory. If the occupancy goes beyond 95%, this parameter is assumed to be breached.
+2. **Node memory limit breach:** The node memory limit is breached if the total allocated node level memory limit for indexing requests is exceeded. At any point the current memory utilization at the node level goes beyond 70% of total assigned node memory for indexing work, this parameter is assumed to be breached.
+
+Breach of primary parameters reflect no actual rejections yet, but triggers an evaluation of the secondary signals.
+
+### Secondary Parameters
+
+These parameters are based on granular, shard level performance such as request throughput and successful completion of requests. While these shard states are updated with every request, they are evaluated only when the primary parameters are breached. We have introduced two secondary parameters:
+
+1. **Throughput:** The framework tracks request latency per unit byte at the shard level and any substantial decrease in the shard throughput against its historic performance is considered a degradation.
+2. **Successful Request:** The framework also tracks successful indexing requests at the shard level and signals if there are large number of pending requests. This is used to detect memory build-ups due to stuck requests proactively and invoke the required backpressure.
+
+Indexing backpressure takes effect when one of the primary parameters and secondary parameters’ thresholds are breached on the node. This results in rejection of new incoming requests/tasks for the affected shards on the node
+The mechanism used by indexing backpressure is fair. The selective rejections of requests are intended only for problematic index or shard, while still allowing others to continue.  Below is the shard state transition diagram which covers these scenarios.
+
+![shard state transition](/assets/media/blog-images/2022-02-04-shard-indexing-backpressure-in-opensearch/shard_state_transition.png){: .img-fluid}
+### **Stats API For Shard Indexing Backpressure**
+
+The performance and rejection statistics collected by the framework are exposed through Rest APIs for users to have added visibility.
+
+
+* `GET /_nodes/_local/stats/shard_indexing_pressure`
+  * Returns the Node level stats for indexing request rejections.
+
+* `GET /_nodes/stats/shard_indexing_pressure`
+  * Returns the Shards level breakup of indexing stats and rejections for every node. This only includes the hot shards i.e. shards which have active writes currently going on them.
```

Comment:
- 'Shards' -> 'shards' (does this need to be plural?)
- 'i.e. shards which have active writes currently going on them' -> '(i.e. shards which have active writes currently going on them)'

Reply:
Agree it can be just ``shard``